linmod/eval.py

```diff
@@ -1,124 +1,210 @@
+        `samples` should have the standard model output format.
+        `data` should have the standard model input format.
+        """
+
+        # Join the forecast samples and raw data dataframes.
+        # Also compute the true proportions from the raw data.
+        self.df = (
+            data.with_columns(
+                phi=(
+                    pl.col("count")
+                    / pl.sum("count").over("fd_offset", "division")
+                ),
+            )
+            .drop("count")
+            # NaN will appear on division-days where none of the lineages
+            # were sampled. We will filter these out.
+            # TODO: Is this good to do?
+            .filter(pl.col("phi").is_not_nan())
+            .join(
+                samples,
+                on=("fd_offset", "division", "lineage"),
+                how="left",
+                suffix="_sampled",
+            )
         )
-    )
-
-    return result
-
 
-def proportions_mean_norm_per_division_day(samples, data, p=1):
-    r"""
-    The expected norm of proportion forecast error for each division-day.
+    def _mean_norm_per_division_day(self, p=1) -> pl.LazyFrame:
+        r"""
+        The expected norm of proportion forecast error for each division-day.
 
-    $E[ || f_{tg} - \phi_{tg} ||_p ]$
+        $E[ || f_{tg} - \phi_{tg} ||_p ]$
 
-    `samples` should have the standard model output format.
-    `data` should have the standard model input format.
-
-    Returns a DataFrame with columns `(division, fd_offset, mean_norm)`.
-    """
-
-    return (
-        _merge_samples_and_data(samples, data)
-        .group_by("fd_offset", "division", "sample_index")
-        .agg(norm=pl_norm(pl.col("phi") - pl.col("phi_sampled"), p))
-        .group_by("fd_offset", "division")
-        .agg(mean_norm=pl.mean("norm"))
-    )
+        Returns a data frame with columns `(division, fd_offset, mean_norm)`.
+        """
 
+        return (
+            self.df.group_by("fd_offset", "division", "sample_index")
+            .agg(norm=pl_norm(pl.col("phi") - pl.col("phi_sampled"), p))
+            .group_by("fd_offset", "division")
+            .agg(mean_norm=pl.mean("norm"))
+        )
 
-def proportions_mean_norm(
-    samples: pl.LazyFrame, data: pl.LazyFrame, p=1
-) -> float:
-    r"""
-    The expected norm of proportion forecast error, summed over all divisions and days.
+    def mean_norm(self, p=1) -> float:
+        r"""
+        The expected norm of proportion forecast error, summed over all divisions
+        and days.
 
-    $\sum_{t, g} E[ || f_{tg} - \phi_{tg} ||_p ]$
-    """
+        $\sum_{t, g} E[ || f_{tg} - \phi_{tg} ||_p ]$
+        """
 
-    return (
-        proportions_mean_norm_per_division_day(samples, data, p=p)
-        .collect()
-        .get_column("mean_norm")
-        .sum()
-    )
+        return (
+            self._mean_norm_per_division_day(p=p)
+            .collect()
+            .get_column("mean_norm")
+            .sum()
```

Comment:
The other way to do this is `self._mean_norm_per_division_day(p=p).select(pl.col('mean_norm').sum()).collect().item()`